ENH: Make swapaxes always return a view. Fixes #5260

jaimefrio opened the pull request.

charris commented on numpy/core/src/multiarray/shape.c:
```diff
@@ -683,25 +672,42 @@ PyArray_SwapAxes(PyArrayObject *ap, int a1, int a2)
+        PyObject *ret;
+
+        Py_INCREF(PyArray_DESCR(ap));
+        ret = PyArray_NewFromDescr(Py_TYPE(ap), PyArray_DESCR(ap), n,
```
Any idea what the last argument does? It isn't documented... OK, looks like it is needed for `__finalize__`, can be `NULL`.

jaimefrio replied:
I used the code in `PyArray_Transpose`, which is the next function after this, as a template. We could actually eliminate this fast path altogether and let `PyArray_Transpose` handle the degenerate case also. There will be some unnecessary overhead, but perhaps simplicity is a better tradeoff than performance here.

charris commented:
Hmm, looks like `PyArray_Transpose` always returns a view and will work if `a1 == a2`, so all that is needed is to avoid returning the original array directly. I wouldn't worry about a fast paths here.

This PR affects the C-API, so that should be documented. I'm inclined to think of this as a buglet fix, as the documentation says it returns a view. It's hard to think of a case where it could cause problems that wasn't due to a user bug.

charris commented:
Apart from that, LGTM ;)

jaimefrio committed:
ENH: Make swapaxes always return a view

swapaxes now returns a view of the input array when the axes to swap
are both the same, not the input array as it used to do. Fixes #5260

jaimefrio committed:
TST: swapaxes tests

jaimefrio commented:
Removed the fast path and added a *C API* remark to the release notes.

charris commented:
Merged, thanks. BTW, I grepped for `PyArray_SetBaseObject` and found 35 spots that may be views being created. Could be worth looking at making a function for that. That would make it easy to add a fast path if we wanted to.

jaimefrio commented:
That would be a new C API function, right?

charris commented:
Could be. It would be interesting to see how common it is in extension code.
